redash/handlers/query_results.py

```diff
@@ -232,6 +233,27 @@ def get(self, query_id=None, query_result_id=None, filetype='json'):
                     abort(404, message='No cached result found for this query.')
 
         if query_result:
+            if models.db.session.query(models.Query, models.DataSource).filter(
+                    models.Query.query_hash == query_result.query_hash,
+                    models.DataSource.type == 'result').first():
+                # Hang on tight, this will be interesting.
+                from redash import extract_table_names
+                for table_name in extract_table_names.extract_tables(query_result.query_text):
+                    # Look for query IDs being accessed.
+                    if table_name.startswith("query_"):
+                        try:
+                            qid = int(table_name.split('_', 1)[1])
+                        except ValueError:
+                            # If it's not "query_NNN" it can't affect our permissions check here.
+                            continue
+                        upstream_q = models.db.session.query(models.Query).filter(models.Query.id == qid).first()
```

Comment:
Let's use the short form here:
```suggestion
                        upstream_q = models.Query.query.filter(models.Query.id == qid).first()
```